content/blog/build-complex-pipelines-with-multiplexer/index.md

```diff
@@ -0,0 +1,347 @@
+
+from haystack import component
+from haystack.dataclasses import Document
+
+@component
+class HallucinationChecker:
+
+    def __init__(self, model: str = "roberta-large-mnli", entailment_threshold: float = 0.5):
+        self.model_name_or_path = model
+        self.entailment_threshold = entailment_threshold
+
+    def warm_up(self):
+        self.tokenizer = AutoTokenizer.from_pretrained(self.model_name_or_path)
+        self.model = AutoModelForSequenceClassification.from_pretrained(pretrained_model_name_or_path=self.model_name_or_path)
+        self.model.to("cuda")
+        id2label = AutoConfig.from_pretrained(self.model_name_or_path).id2label
+        self.labels = [id2label[k].lower() for k in sorted(id2label)]
+
+    @component.output_types(statement=str, supporting_documents=List[Document], hallucination=str, contraddicting_documents=List[Document])
+    def run(self, statement: str, documents: List[Document]):
+        print(f"Checking the statement: '{statement}' against these documents: " + " | ".join([d.content for d in documents]))
+        scores, agg_con, agg_neu, agg_ent = 0, 0, 0, 0
+        entailment_info_batch = self.get_entailment_batch(
+            premise_batch=[doc.content for doc in documents], 
+            hypothesis_batch=[statement] * len(documents)
+        )
+        agg_ent = 0
+        entailing_docs = []
+        contraddicting_docs = []
+        for i, (doc, entailment_info) in enumerate(zip(documents, entailment_info_batch)):
+            if entailment_info["entailment"] > entailment_info["contradiction"]:
+                agg_ent += entailment_info["entailment"]
+                entailing_docs.append(doc)
+                if agg_ent / len(entailing_docs) > self.entailment_threshold:
+                    print("The statement is not a hallucination")
+                    return {"statement": statement, "supporting_documents": entailing_docs}
+            elif entailment_info["contradiction"] > entailment_info["neutral"]:
+                contraddicting_docs.append(doc)
+
+        print("The statement is a hallucination")
+        return {"hallucination": statement, "contraddicting_documents": contraddicting_docs}
+
+    def get_entailment_batch(self, premise_batch: List[str], hypothesis_batch: List[str]):
+        formatted_texts = [
+            f"{premise}{self.tokenizer.sep_token}{hypothesis}"
+            for premise, hypothesis in zip(premise_batch, hypothesis_batch)
+        ]
+        with torch.inference_mode():
+            inputs = self.tokenizer(formatted_texts, return_tensors="pt", padding=True, truncation=True).to("cuda")
+            out = self.model(**inputs)
+            probs_batch = torch.nn.functional.softmax(out.logits, dim=-1).detach().cpu().numpy()
+        return [{k.lower(): v for k, v in zip(self.labels, probs)} for probs in probs_batch]
+```
+
+</details>
+
+With this new component ready we can now start to build the pipeline. 
+
+Our first attempt may look like this:
```

Comment:
```suggestion
With this new component ready we can now start to build the pipeline. The idea is the following:
- The query passes to the retriever, then to the prompt builder, and then to the LLM like in a regular RAG pipeline.
- After the LLM, the reply is given to the `HallucinationChecker` along with the documents returned by the retriever.
- If `HallucinationChecker` decides that the reply is not a hallucination, the pipeline finishes. Otherwise, `HallucinationChecker` sends the LLM output to another prompt builder, that creates a new prompt for the LLM. In this new prompt, the error that the LLM made in the first pass is highlighted and the model is asked to correct it.
- The new prompt is sent to the LLM again and the loop restarts.

Let's implement it. Our first attempt may look like this:
```